File: laboratories/ppo/exercises/lab09/php/src/InputReader.php
<?php

namespace Ppo\Lab09;

use Ppo\Lab09\Exceptions\EmptyStringException;
use Ppo\Lab09\Exceptions\ExitCalledException;
use Ppo\Lab09\Exceptions\ForbiddenCharacterException;
use Ppo\Lab09\Exceptions\MultipleCharactersException;
use Ppo\Lab09\Exceptions\WriteCalledException;

class InputReader
{
    protected string $message = "";

    /**
     * @throws EmptyStringException
     * @throws ExitCalledException
     * @throws ForbiddenCharacterException
     * @throws MultipleCharactersException
     * @throws WriteCalledException
     */
    public function read(): void
    {
        $input = readline("Please type a letter: ");

        if (empty($input)) {
            throw new EmptyStringException();
        }

        if ($input === ":q") {
            throw new ExitCalledException();
        }

        if ($input === ":w") {
            throw new WriteCalledException($this->message);
        }

        if (mb_strlen($input) > 1) {
            throw new MultipleCharactersException();
        }

        if (in_array($input, ["ą", "ć", "ę", "ł", "ń", "ó", "ś", "ź", "ż"])) {
            throw new ForbiddenCharacterException($input);
        }

        $this->message .= strtoupper($input);
    }
}

File: laboratories/ppo/exercises/lab09/python/src/InputReader.py
from src.exceptions.EmptyStringException import EmptyStringException
from src.exceptions.ExitCalledException import ExitCalledException
from src.exceptions.ForbiddenCharacterException import ForbiddenCharacterException
from src.exceptions.MultipleCharactersException import MultipleCharactersException
from src.exceptions.WriteCalledException import WriteCalledException


class InputReader:
    message = ""

    def read(self):
        input_value = input("Please type a letter: ")

        if not input_value:
            raise EmptyStringException()

        if input_value == ":q":
            raise ExitCalledException()

        if input_value == ":w":
            raise WriteCalledException(self.message)

        if len(input_value) > 1:
            raise MultipleCharactersException()

        if input_value in ["ą", "ć", "ę", "ł", "ń", "ó", "ś", "ź", "ż"]:
            raise ForbiddenCharacterException(input_value)

        self.message += input_value.upper()
